### PR TITLE
Fix segmentation fault when assigning to a parenthesized non-tuple expression of a tuple type

### LINKED ISSUE #15075: Segmentation fault when assigning to a parenthesized non-tuple expression of a tuple type
## Environment

- Compiler version: 0.8.25
- Target EVM version (as per compiler settings): None
- Framework/IDE (e.g. Truffle or Remix): None
- EVM execution environment / backend / blockchain client: Command
- Operating system: Linux

## Steps to Reproduce
```solidity
contract A {
    function f ( ) internal virtual {
        ( ( f() ) -= f() ) [ f() ] ;
    }
}
```
Segmentation fault

### PR DESCRIPTION
**Pull request by r0qs**
Fixes https://github.com/ethereum/solidity/issues/15075

### CODE REVIEW

File: test/libsolidity/syntaxTests/tupleAssignments/empty_tuple_lhs_resolved.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+contract C {
+	function f() public {
+		(f()) = ();
```
Comment (cameel) [inline test/libsolidity/syntaxTests/tupleAssignments/empty_tuple_lhs_resolved.sol]: Another test case that reproduces this and does not even involve an empty tuple:

```solidity
    function g() internal returns (uint, uint) {}

    function f() public {
        (g()) = (1, 1);
    }
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -116,7 +116,7 @@ void TypeChecker::checkDoubleStorageAssignment(Assignment const& _assignment)
 		TupleType const& lhsType = dynamic_cast<TupleType const&>(*type(_lhs));
 		TupleExpression const* lhsResolved = dynamic_cast<TupleExpression const*>(resolveOuterUnaryTuples(&_lhs));
 
-		if (lhsType.components().size() != _rhs.components().size() || lhsResolved->components().size() != _rhs.components().size())
+		if (!lhsResolved || lhsType.components().size() != _rhs.components().size() || lhsResolved->components().size() != _rhs.components().size())
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: I think this would make more sense:

```suggestion
		if (lhsType.components().size() != _rhs.components().size() || (lhsResolved && lhsResolved->components().size() != _rhs.components().size()))
```
Your version makes an implicit assumption that when `lhsResolved` is null, the code cannot be correct. While this assumption *is* true, that's only really because currently you cannot return tuple L-values from functions. We could easily avoid this hidden assumption.

Moreover, the second part of the condition seems superfluous to me. I think that non-null `lhsResolved` will always match the number of components in `lhsType`, so I'd actually change it to this:

```suggestion
		solAssert(!lhsResolved || lhsResolved->components().size() == lhsType.components().size());
		if (lhsType.components().size() != _rhs.components().size())
```

Comment (cameel) [reply libsolidity/analysis/TypeChecker.cpp]: Actually, let me explain it in more detail, because there are a few gotchas about our type system that might not be obvious and even knowing these details it took me a while to think through the possible cases here.

We're dealing with `TupleType` and `TupleExpression` here and, perhaps counterintuitively, one does not imply the other. Something can be of `TupleType` even if it's not a `TupleExpression`. For example a `FunctionCall` expression `f()` will get a tuple type when the function returns multiple values or zero values. And conversely, something can be a `TupleExpression` even if it's not of a `TupleType`. For example `()`, `(1)` and `(1, 1)` are all tuple expressions, but only the first and the last are of tuple type. `(1)` is a parenthesized rational number. Parenthesized expressions are the only special case - in that case the type comes from the content. So e.g. `(1)`, `((((1, 2, 3)))` and `((f())` have the same types as `1`, `(1, 2, 3)` and `f()`, respectively. Note that an empty tuple is of `TupleType` even though it has no content.

So now, `TypeChecker::checkDoubleStorageAssignment()` assumes that both sides of the assignment are of `TupleType` (I think it should really have explicit asserts against that). Otherwise it won't get called. This filters out some possibilities for LHS (e.g. cases like `(x) = (1)` that would otherwise fail the condition here). We're left with:
1. Not a single element tuple expression. The type comes directly from that expression and has the same number of components.
    ```solidity
    function f() public {
        (uint x, uint y) = (1, 1); // OK
    }
    ```
1. Single element tuple expression containing a non-single element tuple expression. The type comes directly from that nested expression and has the same number of components.
    ```solidity
    function f() public {
        ((x.push(), x.push())) = (1, 1); // OK
    }
    ```
1. Single element tuple expression containing something else of tuple type (e.g. a function returning a tuple). In that case `lhsResolved` is null.
    ```solidity
    function f() public returns (uint, uint) {
        (f()) = (1, 1); // Error: Expression has to be an lvalue.
    }
    ```

For that reason checking the number of components in `lhsResolved` is redundant. It either follows from the number of components in the type or is irrelevant.

**Review by cameel — COMMENTED** [review]
The new condition is technically incorrect (or rather makes unnecessary assumptions), though cases that it would process incorrectly are currently impossible in the language, so the code is not buggy despite this. I think it would still be better to correct it before we merge the PR.

File: test/libsolidity/syntaxTests/tupleAssignments/empty_tuple_lhs_resolved.sol (hunk 1 above)
Comment (r0qs) [reply test/libsolidity/syntaxTests/tupleAssignments/empty_tuple_lhs_resolved.sol]: I added this to the test cases.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (r0qs) [reply libsolidity/analysis/TypeChecker.cpp]: Right. Thanks for the explanation, I wasn't aware of such distinction ;)
I applied the suggestion and also your examples as test cases.

File: Changelog.md [hunk 3]
```diff
@@ -16,6 +16,7 @@ Bugfixes:
  * SMTChecker: Fix internal error when using bitwise operators with an array element as argument.
  * Standard JSON Interface: Fix ICE when the optimizer is disabled and an empty/blank string is used for ``optimizerSteps`` sequence.
  * StaticAnalyzer: Only raise a compile time error for division and modulo by zero when it's between literals.
+ * TypeChecker: Fix segmentation fault when assigning to complex left side tuple expression that resolves to empty.
```
Comment (cameel) [inline Changelog.md]: This changelog entry will be hard to understand to users. I don't think it's clear that "resolving" here means removing useless parentheses. It's just our internal terminology used in the implementation, not official language term.

It's also incorrect because it's not about empty tuples after all.

I'd use this:
```suggestion
 * TypeChecker: Fix compiler crash when the left-hand side of an assignment was a parenthesized non-tuple expression of a tuple type.
```
I guess it's still hard to understand without knowing what I wrote in that long comment (https://github.com/ethereum/solidity/pull/15084#discussion_r1594499648), but at least should be correct :)

File: test/libsolidity/syntaxTests/tupleAssignments/empty_tuples_lhs_resolved.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/tupleAssignments/empty_tuples_lhs_resolved.sol:1]: This name is no longer correct. Maybe call it `assignments_to_tuple_and_non_tuple_expressions_of_tuple_types`?

Also, probably would not hurt to add a comment for each case, remarking where we have tuple expressions, where non-tuple expressions and where tuple-expressions wrapped in parentheses. Also that all of these things are still of tuple types here.

Alternatively, each case could also be a separate test with an appropriate name.

**Review by cameel — DISMISSED** [review]
I'm going to tentatively approve it since my only remaining problem here is wording and comments, but would still be good if those could be corrected before we merge.

File: test/libsolidity/syntaxTests/tupleAssignments/empty_tuples_lhs_resolved.sol
Comment (r0qs) [reply test/libsolidity/syntaxTests/tupleAssignments/empty_tuples_lhs_resolved.sol:1]: Right. I added some comments. Let me know if they are correct or I got something wrong ;)

File: test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol [hunk 4]
```diff
@@ -0,0 +1,34 @@
+contract C {
+	uint[] public array;
+
+	function f() public {
+		(f()) = (); // Function call resolving to an empty tuple expression of type tuple(). Error
+	}
+
+	function g() public {
+		(revert()) = (); // Buildin function call resolving to an empty tuple expression of type tuple(). Error
```
Comment (cameel) [inline test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol]:
```suggestion
		(revert()) = (); // Builtin function call resolving to an empty tuple expression of type tuple(). Error
```

File: test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol [hunk 5]
```diff
@@ -0,0 +1,34 @@
+contract C {
+	uint[] public array;
+
+	function f() public {
+		(f()) = (); // Function call resolving to an empty tuple expression of type tuple(). Error
+	}
```
Comment (cameel) [inline test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol]: Well, this is still not quite right. The point of these comments was to point out where the whole expression on the left-hand side is a tuple expression. For example:

```suggestion
	function f() public {
		// (f()) is not a tuple expression, but its value is a tuple.
		(f()) = ();
	}
```

Comment (cameel) [reply test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol]: This was just an example to show what I meant. But all the comments in the test have this problem, not just this one.

Comment (r0qs) [reply test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol]: I know, I updated the others as well ;)

File: test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+		(j()) = (1, 1);
+	}
+
+	function m() public {
+		// (uint x, uint y) is a tuple expression, and its value is a tuple (uint, uint).
+		(uint x, uint y) = (1, 1);
+	}
+
+	function n() public {
+		// ((array.push(), array.push())) is a tuple expression of a tuple expression, and its value is a tuple (pointer uint, pointer uint).
```
Comment (cameel) [inline test/libsolidity/syntaxTests/tupleAssignments/assignments_to_tuple_and_non_tuple_expressions_of_tuple_types.sol]:
```suggestion
		// ((array.push(), array.push())) is not a tuple expression, but contains a tuple expression, and the value of both is a tuple (pointer uint, pointer uint).
```

**Review by cameel — APPROVED** [review]
